### PR TITLE
search contexts: allow creation of dynamic search contexts with repo:has, repo:has.tag, and repo:has.key

### PR DESCRIPTION
**Pull request by tbliu98**
Stacked on #44873

Addresses #43958

Allow the use of `repo:has`, `repo:has.tag` and `repo:has.key` in query-based search contexts. This way, users can create a search context like `repo:has.tag(my-team-name)` once to encompass all current and future repos that are tagged with "my-team-name".

## Test plan
Manually test
Added unit tests

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.01% (+0.16 kb) | 0.01% (+0.88 kb) | 0.01% (+0.72 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/tl/repo-kvps-search-contexts/compare-report.html;tab=live_object) for a full comparison between the commits 6460621 and f8985dd or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Comment by camdencheek** [comment]
For some reason, I was thinking this would be more complicated than this. I thought we did some magic like cache the repos resolved by a dynamic context. I don't see anything like that though

**Review by camdencheek — APPROVED** [review]
Approved pending the comment about the return

File: internal/search/searchcontexts/search_contexts.go [hunk 1]
```diff
@@ -207,9 +207,12 @@ func validateSearchContextQuery(contextQuery string) error {
...
 			if a.Labels.IsSet(query.IsPredicate) {
-				errs = errors.Append(errs,
-					errors.Errorf("unsupported repo field predicate in search context query: %q", value))
-				return
+				predName, _ := query.ParseAsPredicate(value)
+				if predName != "has" && predName != "has.tag" && predName != "has.key" {
+					errs = errors.Append(errs,
+						errors.Errorf("unsupported repo field predicate in search context query: %q", value))
+					return
+				}
```
[the remark this answers is not in the document]
Comment (tbliu98) [reply internal/search/searchcontexts/search_contexts.go:215]: Oh yes, thanks for the catch

**Pull request merged**